README.md
![APRSMessenger banner](https://raw.githubusercontent.com/rhymeswithmogul/APRSMessenger/main/.github/repository-open-graph-template.png)

# APRSMessenger

[![PowerShell Gallery Version](https://img.shields.io/powershellgallery/v/APRSMessenger)
 ![PowerShell Gallery](https://img.shields.io/powershellgallery/p/APRSMessenger)
 ![Download from PowerShell Gallery](https://img.shields.io/powershellgallery/dt/APRSMessenger)
](https://www.powershellgallery.com/packages/APRSMessenger/)
![GitHub Workflow Status](https://img.shields.io/github/actions/workflow/status/rhymeswithmogul/APRSMessenger/run-pester-tests.yaml)
[![License: AGPL v3](https://img.shields.io/badge/License-AGPL_v3-blue.svg)](https://www.gnu.org/licenses/agpl-3.0)
[![Run Pester tests](https://github.com/rhymeswithmogul/APRSMessenger/actions/workflows/run-pester-tests.yaml/badge.svg)](https://github.com/rhymeswithmogul/APRSMessenger/actions/workflows/run-pester-tests.yaml)

A PowerShell module to send messages, bulletins, and announcements through the APRS-IS network.

## Messaging
You can use this module to create APRS messages.  You will need to specify a server name in order to send to APRS-IS.  By default, the username and password are derived from your callsign, and the default port (TCP 14580) is used.

```powershell
Send-APRSMessage `
	-From N0CALL -To FR1END `
	-Message "Hey, how's it going?" `
	-Server noam.aprs2.net
```

Another example:  to check into the [APRS Thursday net](https://aprsph.net/aprsthursday/), you can do something like this.  Replace `AB12cd` with your location and/or grid square.  Optionally, include `-Latitude` and `-Longitude`, and supported APRS receivers can plot your message on a map.

```powershell
Send-APRSMessage `
	-From N0CALL -To ANSRVR `
	-Message "CQ HOTG Happy #APRSThursday from AB12cd" `
	-Latitude 12.34  -Longitude -56.7890
	-Server rotate.aprs2.net
```

If you do not specify a server or port, the APRS packet will be printed to the screen.  This might be useful if you want to transmit it via some other means, such as through a locally-attached radio.

```powershell
Send-APRSMessage -From N0CALL -To RFL1NK -Message "Hello!"
```
```
N0CALL>RFL1NK,TCPIP*::RFL1NK   :Hello!
```

### Message Acknowledgements
To request a message acknowledgement from the recipient, specify a message identifer with the `-Acknowledge` parameter.

```powershell
Send-APRSMessage `
	-From N0CALL -To FR1END `
	-Message "Let me know if you receive this." `
	-Acknowledgement 69
```
```
N0CALL>FR1END,TCPIP*::FR1END   :Let me know if you receive this.{69
```

Note that APRS, PowerShell, nor this module can acknowledge messages; this must be supported by the receiving hardware (usually a ham radio or something).  That is, this module can only send messages;  you will need some other hardware to listen for the acknowledgement or rejection.


## Announcements
To send an APRS announcement to anyone who might be listening for one:
```powershell
Send-APRSAnnouncement `
	-From GR8CLUB `
	-AnnouncementID 'R' `
	-Message "We will be providing communications for the road race on Sunday. Meet at the starting line." `
	-Latitude -12.3456 -Longitude 78.90123
```

## Bulletins
To send an APRS bulletin to anyone who might be listening for one:
```powershell
Send-APRSBulletin `
	-From GR8CLUB `
	-BulletinID 2 `
	-Message "Road race volunteers, listen to bulletins with group name RACE."
```
```powershell
Send-APRSBulletin `
	-From GR8CLUB-13 `
	-BulletinID 1 `
	-Message "Rain in the forecast later. Should not affect road race."
```

## Group Bulletins
To send an APRS bulletin to a certain group who might be listening for one:
```powershell
Send-APRSGroupBulletin `
	-From GR8CLUB `
	-BulletinID 2 -GroupName "RACE" `
	-Message "Road race traffic net on 146.55 MHz simplex. W1DNS handling emcomm."
```
```powershell
Send-APRSGroupBulletin `
	-From GR8CLUB-13 `
	-BulletinID 1 -GroupName "WX" `
	-Message "There will be a SKYWARN net on the repeater tonight at 8:00 PM."
```

## Example: Putting It All Together
Here is an example where you could use APRS messaging to manage a club's weather station.  Let's assume that a hypothetical `Get-WeatherStationData` cmdlet returns all of your collected weather data at any given moment, and the following script runs once per minute.  For brevity, I am omitting APRS-IS connection data.

```powershell
#Requires -Module MyCoolWeatherStationModule
$wxdata = (Get-MyCoolWeatherStation).Data
$now    = (Get-Date)

# Send W1DNS an alert if the battery needs to be replaced.
If ($wxdata.BatteryVoltage -lt 11  -and  $now.Minute -eq 0) {
	Send-APRSMessage -From "GR8CLUB-13" -To "W1DNS" `
		-Message "Please replace the battery in the weather station $($wxdata.Name)."
}

# Report if there's been a lightning strike detected.
If ($wxdata.LightningStrikes -gt 0) {
	Send-APRSGroupBulletin -From "GR8CLUB-13" -AnnouncementID "L" `
		-Message "A lightning strike was detected.  Information: $($wxdata.LightningStrikes[0])"
	Clear-RecordedLightningStrikes
}

# Broadcast a weather report at thirty past the hour.
If ($now.Minute -eq 30)
{
	$temp = $wxdata.Temperature
	$cond = $wxdata.Conditions
	$rain = $wxdata.RainfallPastHour
	$wspd = $wxdata.WindSpeed
	$wdir = $wxdata.WindDirection
	Send-APRSGroupBulletin -From "GR8CLUB-13" -AnnouncementID "W" `
		-Message "Weather report: $temp° and $cond, $rain`" rainfall, wind $wspd MPH from $wdir."
}

# At midnight and noon, broadcast announcements for a club's weekly SKYWARN net.
# On Sundays, send out a bulletin announcing tonight's net.
If ($now.Hour % 12 -eq 0  -and  $now.Minute -eq 0) {
	If ($now.DayOfWeek -ne "Sunday") {
		Send-APRSAnnouncement -From "GR8CLUB-13" -AnnouncementID "S" `
			-Message "There will be a SKYWARN net on GR8CLUB's repeater Sunday at 8:00 PM."
	} Else {
		Send-APRSBulletin -From "GR8CLUB-13" -BulletinID 1 `
			-Message "There is a SKYWARN net on GR8CLUB's repeater tonight at 8:00 PM."
	}
}
```

***

## Supplemental Reading

### What Is APRS?
You don't know?  From [the official site](http://aprs.org):
> The Automatic Packet Reporting System was designed to support rapid, reliable exchange of information for local, tactical real-time information, events or nets.

and

> APRS is digital communications information channel for Ham radio. As a single national channel[…], it gives the mobile ham a place to monitor for 10 to 30 minutes in any area, at any time to capture what is happening in ham radio in the surrounding area. Announcements, Bulletins, Messages, Alerts, Weather, and of course a map of all this activity including objects, frequencies, satellites, nets, meetings, Hamfests, etc. The APRS network has grown to most countries with strong Amateur Radio populations. 


### Legal Warning
To use this app, you must be a licensed amateur radio operator.  [Getting your license is easy.](https://hamstudy.org/)

Like it says in this app's license: this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the [GNU Affero General Public License 3.0](https://www.gnu.org/licenses/agpl-3.0.html) (or any later version0 for more details. As such, you and you alone are solely responsible for using this app. Please do not use this app for evil. Don't make me regret writing this app.

This project is not affiliated with nor endorsed by the APRS Tier 2 Network, the APRS Working Group, aprs.org, the family of Bob Bruninga -- or anyone, for that matter.

73, W1DNS

NEWS.md
# News for APRSMessenger

## Version 1.1.0 (April 20, 2023)
We've added the ability to include position reports in your messages, bulletins, and announcements.  Simply use the `-Latitude` and `-Longitude` parameters to include them.  Then, they will show up on sites like [APRS.fi](https://APRS.fi).

## Version 1.0.4 (March 29, 2023)
In this version, we've signed the PowerShell Gallery version of this code so that it can load on platforms that require code signing in the execution policy.  Some Pester tests were also added to ensure the module is well-formed.

## Version 1.0.2 (March 21, 2023)
In this version, we've fixed bugs related to group bulletins not being received correctly, message acknowledgements not working, and printing packets to the output stream.

We've also included Pester testing to ensure that this app works as intended.

Finally, this version contains a catalog file to verify its authenticity, and to help it run when execution policies require signed code. 

## Version 1.0.1 (March 20, 2023)
Fixed a very embarrassing bug where APRS messages wouldn't be sent to the APRS-IS network.  Oops.

## Version 1.0.0 (March 19, 2023)
Initial release.

release/New-Release.ps1
# This file is part of APRSMessenger.
#
# APRSMessenger is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# APRSMessenger is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License
# along with APRSMessenger. If not, see <https://www.gnu.org/licenses/>. 

#Requires -Module Microsoft.PowerShell.Security, PSScriptAnalyzer, @{ModuleName='Pester';ModuleVersion='5.0.0'}

If (-Not $IsWindows) {
	Throw [PlatformNotSupportedException]::new('This script requires Microsoft Windows.')
}

#region Create a build directory.
# We will copy all important files to a temporary folder.  We're using
# New-Temporary file only to create a filename that the runtime guarantees to
# be unique.
$TempFile = New-TemporaryFile
Remove-Item -Path $TempFile
$DestinationPath = (Join-Path -Path $env:Temp -ChildPath $TempFile.Name)

Write-Output "Copying module to $DestinationPath"
New-Item -Path $DestinationPath -ItemType Directory -ErrorAction Stop
Copy-Item -Path . -Destination $DestinationPath -Recurse -Exclude @(
	'.git*',		# This can be retrieved from GitHub.
	'coverage.xml',	# junk
	'man',			# Get-Help should be used instead.
	'release'		# You don't need this script.  Only I do.

	# We only need the PNG icon.
	'icon/APRSMessenger.avif',
	'icon/APRSMessenger.jxl',
	'icon/APRSMessenger.webp',
	'icon/APRSMessenger.xcf',
	'icon/README.md'
)
Push-Location -Path (Join-Path -Path $DestinationPath -ChildPath 'APRSMessenger')
#endregion

#region Sign all script files.
# This portion of the script signs all files with my code signing certificate.
# Since the command's default parameters are defined in my shell, and my private
# key requires protection, there are no secrets to hide in this script.  This
# will silently fail on all other computers except mine.
Write-Output "Signing all script files"
Get-ChildItem -Recurse -Include @('*.ps1','*.ps?1') | ForEach-Object {
	Set-AuthenticodeSignature $_ | Format-Table -AutoSize
}

Write-Output "Generating catalog"
New-FileCatalog -Path . -CatalogFilePath APRSMessenger.cat -CatalogVersion 2.0
Set-AuthenticodeSignature 'APRSMessenger.cat'
#endregion

#region Invoke PSScriptAnalyzer.
Write-Output "Calling PSScriptAnalyzer with Gallery settings"
$analysis = Invoke-ScriptAnalyzer -Path . -Recurse -Settings PSGallery
Write-Output $analysis
If ($analysis.Count -gt 0)
{
	Throw 'Please correct PSScriptAnalyzer errors and try again.'
}
#endregion

#region Run Pester tests.
Write-Output "Running Pester tests"
Invoke-Pester
#endregion

Start-Process -FilePath 'C:\Windows\Explorer.exe' -ArgumentList $DestinationPath
Pop-Location


APRSMessenger.Tests.ps1
# This file is part of APRSMessenger.
#
# APRSMessenger is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# APRSMessenger is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License
# along with APRSMessenger. If not, see <https://www.gnu.org/licenses/>.

[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseDeclaredVarsMoreThanAssignments', 'ExampleAnnouncement',	Justification='Variable is used in another scope.')]
[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseDeclaredVarsMoreThanAssignments', 'ExampleBulletin',		Justification='Variable is used in another scope.')]
[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseDeclaredVarsMoreThanAssignments', 'ExampleGroupBulletin',	Justification='Variable is used in another scope.')]
[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseDeclaredVarsMoreThanAssignments', 'ExampleEmail',			Justification='Variable is used in another scope.')]
[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseDeclaredVarsMoreThanAssignments', 'ModuleHelpFile',		Justification='Variable is used in another scope.')]
[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseDeclaredVarsMoreThanAssignments', 'psm1File',				Justification='Variable is used in another scope.')]
[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseDeclaredVarsMoreThanAssignments', 'ThisIsATest',			Justification='Variable is used in another scope.')]
[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseDeclaredVarsMoreThanAssignments', 'WeatherReport',			Justification='Variable is used in another scope.')]
Param()

BeforeAll {
	Import-Module -Name (Join-Path -Path '.' -ChildPath 'APRSMessenger.psd1')

	$ThisIsATest          = 'This is a test.'
	$ExampleAnnouncement  = 'My St Helen digi will be QRT this weekend'
	$ExampleBulletin      = 'Snow expected in Tampa RSN'
	$ExampleGroupBulletin = 'Stand by your snowplows'
	$ExampleEmail         = 'msproul@ap.org Test email'
	$WeatherReport        = 'This is a weather report.'
}

Context 'Validate the module files' {
	BeforeAll {
		$psm1File       = Join-Path -Path 'src'   -ChildPath 'APRSMessenger.psm1'
		$ModuleHelpFile = Join-Path -Path 'en-US' -ChildPath 'APRSMessenger-help.xml'
	}
	It 'has a module manifest' {
		'APRSMessenger.psd1' | Should -Exist
	}
	It 'has a root module' {
		$psm1File | Should -Exist
	}
	It 'has a valid root module' {
		$code = Get-Content -Path $psm1File -ErrorAction Stop
		$errors = $null
		$null = [Management.Automation.PSParser]::Tokenize($code, [ref]$errors)
		$errors.Count | Should -Be 0
	}
	It 'has a conceptual help file' {
		Join-Path -Path 'en-US' -ChildPath 'about_APRSMessenger.help.txt' | Should -Exist
	}
	It 'has a module help file' {
		$ModuleHelpFile | Should -Exist
	}
	It 'has a valid module help file' {
		$code = [Xml](Get-Content -Path $ModuleHelpFile -ErrorAction Stop)
		$code.Count | Should -Be 1
	}
}

Describe 'Get-APRSISPasscode' {
	It "Returns <expected> for callsign <name>" -TestCases @(
		@{Name='N0CALL';	Expected=13023},
		@{Name='N0CALL-13';	Expected=13023},
		@{Name='W1APR';		Expected=14211},
		@{Name='W1APR-3';	Expected=14211}
	) {
		Get-APRSISPasscode $name | Should -Be $expected
	}
}

Describe 'Send-APRSMessage' {
	It "Generates a valid message from N0CALL to FR1END with content `"$ThisIsATest`"" {
		Send-APRSMessage -From 'N0CALL' -To 'FR1END' -Message $ThisIsATest `
			| Should -Be "N0CALL>FR1END,TCPIP*::FR1END   :$ThisIsATest"
	}
	It 'Generates the first example message from page 71 of the APRS 1.01 specification document.' {
		Send-APRSMessage -From 'N0CALL' -To 'WU2Z' -Message 'Testing' `
			| Should -Be 'N0CALL>WU2Z,TCPIP*::WU2Z     :Testing'
	}
	It 'Generates the second example message from page 71 of the APRS 1.01 specification document.' {
		Send-APRSMessage -From 'N0CALL' -To 'WU2Z' -Message 'Testing' -Acknowledge 003 `
			| Should -Be 'N0CALL>WU2Z,TCPIP*::WU2Z     :Testing{003'
	}
	It 'Generates the third example message from page 71 of the APRS 1.01 specification document.' {
		Send-APRSMessage -From 'N0CALL' -To 'EMAIL' -Message $ExampleEmail `
			| Should -Be "N0CALL>EMAIL,TCPIP*::EMAIL    :$ExampleEmail"
	}
	It 'Creates a valid message with position information.' {
		Send-APRSMessage -From 'N0CALL' -To 'PESTR' -Message 'Position test' -Latitude 12.345 -Longitude -67.890 `
			| Should -Be "N0CALL>PESTR,TCPIP*:=1220.42N/06753.24W`"PESTR    :Position test"
	}
}

Describe 'Send-APRSAnnouncement' {
	It "Generates a valid announcement from N0CALL with identifier T saying `"$ThisIsATest`"." {
		Send-APRSAnnouncement -From 'N0CALL' -AnnouncementID 'T' -Message $ThisIsATest `
			| Should -Be "N0CALL>BLNT,TCPIP*::BLNT     :$ThisIsATest"
	}
	It "Generates a valid announcement from N0CALL-13 with identifier W saying `"$WeatherReport`"" {
		Send-APRSAnnouncement -From 'N0CALL-13' -AnnouncementID 'W' -Message $WeatherReport `
			| Should -Be "N0CALL-13>BLNW,TCPIP*::BLNW     :$WeatherReport"
	}
	It 'Generates the example announcement from page 73 of the APRS 1.01 specification.' {
		Send-APRSAnnouncement -From 'N0CALL' -AnnouncementID 'Q' -Message $ExampleAnnouncement `
			| Should -Be "N0CALL>BLNQ,TCPIP*::BLNQ     :$ExampleAnnouncement"
	}
	It 'Creates a valid announcement with position information.' {
		Send-APRSAnnouncement -From 'N0CALL' -AnnouncementID 'P' -Message 'Position test' -Latitude 12.345 -Longitude -67.890 `
			| Should -Be "N0CALL>BLNP,TCPIP*:=1220.42N/06753.24W`"BLNP     :Position test"
	}
}

Describe 'Send-APRSBulletin' {
	It "Generates a valid bulletin from N0CALL with identifier 1 saying `"$ThisIsATest`"." {
		Send-APRSBulletin -From 'N0CALL' -BulletinID '1' -Message $ThisIsATest `
			| Should -Be "N0CALL>BLN1,TCPIP*::BLN1     :$ThisIsATest"
	}
	It 'Generates the example general bulletin from page 73 of the APRS 1.01 specification document.' {
		Send-APRSBulletin -From 'N0CALL' -BulletinID 3 -Message $ExampleBulletin `
			| Should -Be "N0CALL>BLN3,TCPIP*::BLN3     :$ExampleBulletin"
	}
	It 'Generates a valid bulletin with position information.' {
		Send-APRSBulletin -From 'N0CALL' -BulletinID '9' -Message 'Position test' -Longitude 123.45678 -Latitude 90 `
			| Should -Be "N0CALL>BLN9,TCPIP*:=9000.00N/12327.24E`"BLN9     :Position test"
	}
}

Describe 'Send-APRSGroupBulletin' {
	It "Generates a valid bulletin from N0CALL with identifier 1 to group PESTR saying `"$ThisIsATest`"." {
		Send-APRSGroupBulletin -From 'N0CALL' -BulletinID '1' -GroupName 'PESTR' -Message $ThisIsATest `
			| Should -Be "N0CALL>BLN1,TCPIP*::BLN1PESTR:$ThisIsATest"
	}
	It 'Generates the example group bulletin from page 74 of the APRS 1.01 specification.' {
		Send-APRSGroupBulletin -From 'N0CALL' -BulletinID '4' -GroupName 'WX' -Message $ExampleGroupBulletin `
			| Should -Be "N0CALL>BLN4,TCPIP*::BLN4WX   :$ExampleGroupBulletin"
	}
	It 'Generates a valid group bulletin with position information.' {
		Send-APRSGroupBulletin -From 'N0CALL' -BulletinID '9' -GroupName 'PESTR' -Message 'Position test' -Latitude 12.345 -Longitude -67.890 `
			| Should -Be "N0CALL>BLN9,TCPIP*:=1220.42N/06753.24W`"BLN9PESTR:Position test"
	}
}

AfterAll {
	Remove-Module -Name 'APRSMessenger'
}
